### PR TITLE
Added support for giving access on connections to users

### PR DESCRIPTION
**Pull request by koszti**
This PR adds the admin functionality to give access on connections to specific users.

Screenshot 2019-12-30 at 14 15 42

**Creating access on specific connection to users.**
The duration is optional, if not set then a permanent access will be created:
Screenshot 2019-12-30 at 14 11 14

**List users with access to specific connections(s).**
Admins optionally can expire active accesses:
Screenshot 2019-12-30 at 14 09 16

### CODE REVIEW

**Comment by rickbergfalk** [comment]
This work looks great @koszti -- unfortunately I'm unsure if it is a fit for the project at this time. 

Today connections are a "global" resource of sorts available to all users (even editor roles). If I'm understanding things correctly, this would remove access from editors unless explicitly provided right?

**Comment by koszti** [comment]
@rickbergfalk , we'd like to use this project but creating temporary access is a must for us. We were thinking to fork but always prefer contributing so I raised this PR in the hope it will be useful for others as well. You're right, the current changes would remove access from editors unless explicitly provided, and I have to agree that is not great and not in sync with the current behaviour.

What do you think, would that make sense creating a **globalResourceAccess** (Env Var: `GLOBAL_RESOURCE_ACCESS` config variable with default `True` ? So nothing in this PR would be visible and active unless somebody explicitly define it in the config?

I'd still prefer contributing if there is any chance and not creating a fork so please tell and suggest anything that would help adding this functionality to the project.

**Comment by rickbergfalk** [comment]
@koszti completely understand and appreciate the desire to contribute back! I've been thinking about this quite a bit today and I believe this would help the use case identified in #499.

I think adding a global variable to turn this feature on makes sense.

Another option to potentially consider -- what do you think about having some sort of special "everyone" user, and defaulting connections to everyone for an indefinite period of time? 

@gonace would be interested on your thoughts on this as well

**Comment by koszti** [comment]
Having an "everyone" virtual user makes lot of sense to me. What if we combine it with an optional global variable in config to work like that:

* By default every user can use every connection for an indefinite period of time. In the UI the Connection Access dropdown will have an extra "everyone" option. In this case we can still expire and re-create the access per connection whenever it's required.
* Adding extra **allowConnectionAccessToEveryone** (Env var `SQLPAD_ALLOW_CONNECTION_ACCESS_TO_EVERYONE`) with Default value `true`. Users can set this to `false` so they can decide to give access automatically on new connections to "everyone" or not.

**Comment by rickbergfalk** [comment]
Sounds great -- I'm on board with that approach.

**Comment by kclotfelter** [comment]
@koszti What would you think of disabling the duration minimum? For my team, the ability to limit connections to specific users is important but access should essentially be granted indefinitely. 

Changing the minimum from 900 to 0 in server/models/connectionAccesses.js allowed me to set access duration to 0 within the Connection Accesses pane, providing an expiry date in the year 2099.

image

Having this feature in SQLPad is really a game-changer, but I wanted to throw this out there in case others have similar use-cases.

**Comment by koszti** [comment]
@rickbergfalk , I added some commits addressing the items that we discussed.

@kclotfelter , the duration field is optional. If you leave it empty you will grant indefinite access to the connection. In your case you set the duration to 0, which actually also means indefinite access until year 2099. At the beginning I didn't find useful giving temp accesses to anybody for less than 15 minutes so I set it to minimum 900 secs. What do you think? Does it make sense or we should still remove the minimum 900 validation rule and allow very short accesses as well?


**CHANGES BY THE LAST COMMITS**

1) **Default connection accesses**
A connection access created automatically at startup to give access on every connection to every user:  
Screenshot 2020-01-02 at 16 20 37

This behaviour can be disabled by setting the `SQLPAD_ALLOW_CONNECTION_ACCESS_TO_EVERYONE` to `false` at startup

2) **Virtual groups for connections and users** 
`Every Connection` and `Every User` is now in the dropdowns to optionally give more flexible accesses
Screenshot 2020-01-02 at 16 26 06

**Comment by kclotfelter** [comment]
@koszti Oh, I gotcha; that makes perfect sense. I just didn't realize it would default to indefinite when left blank. Would it make sense to change the verbiage prior to setting the duration as well? Something as simple as _(optional)_ would probably do the trick.

image

Also, I **love** the virtual groups addition. 

Huge thanks for your work on this, as well as @rickbergfalk and everyone who's contributed to what SQLPad is today.

**Comment by koszti** [comment]
@kclotfelter committed 5fb330288709441076566c8d07cdb204fcb00042 that does the trick on the label

**Comment by gonace** [comment]
This might just be me, but I kind of like `Everyone` more than `Every User`, any thoughts on that @rickbergfalk and @koszti?

File: client/src/connectionAccesses/ConnectionAccessForm.js [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        </HorizontalFormItem>
+        <br />
+        <div
+          style={{
+            borderTop: '1px solid #e8e8e8',
+            paddingTop: '22px',
+            textAlign: 'right'
+          }}
+        >
+          <Button
```
[the remark this answers is not in the document]
Comment (koszti) [reply client/src/connectionAccesses/ConnectionAccessForm.js]: fixed by 7a0f963dba0d98eab638c3a911e7e1fab72e0cad

File: client/src/connectionAccesses/ConnectionAccessForm.js [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+          <option key={user._id} value={user._id}>
+            {user.email}
+          </option>
+        )
+      );
+  }
+
+  return (
+    <div style={{ height: '100%' }}>
+      <form
```
[the remark this answers is not in the document]
Comment (koszti) [reply client/src/connectionAccesses/ConnectionAccessForm.js]: fixed by 7a0f963dba0d98eab638c3a911e7e1fab72e0cad

**Comment by koszti** [comment]
@gonace, you're right. Originally we discussed to use `Everyone` but when I added `Every Connection` I was trying to keep the naming in sync and ended up with `Every User`.

Actually I don't mind and if people prefer `Everyone` then I'm happy to change it. @rickbergfalk what do you think?

**Comment by rickbergfalk** [comment]
I'm in support of everyone as well

**Comment by koszti** [comment]
thanks for the feedback. https://github.com/rickbergfalk/sqlpad/pull/502/commits/48d73fbd5edd87d373ff9a190c2237d03a26fd23 commit renames `Every` User to `Everyone`

**Comment by rickbergfalk** [comment]
Thanks for the flexibility and contribution @koszti!

**Pull request merged**